service/mapper/transitions.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package mapper
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/models/index"
+)
+
+type TransitionFunc func(*State) error
+
+func Bootstrap(root Root, load Loader) TransitionFunc {
+	return func(s *State) error {
+
+		// Bootstrapping should only happen when the state is empty.
+		if !Empty(s) {
+			return fmt.Errorf("invalid state for bootstrap")
+		}
+
+		// We always need at least one step in our forest, which is used as the
+		// stopping point when indexing the payloads since the last finalized
+		// block. We thus introduce an empty tree, with no paths and an
+		// irrelevant previous commit.
+		empty := trie.NewEmptyMTrie()
+		ok := s.forest.Save(flow.DummyStateCommitment, empty, nil)
+		if !ok {
+			return fmt.Errorf("could not save empty tree")
+		}
```

Comment:
So this should only happen if there is already a value at `state.trees[flow.DummyStateCommitment]` which makes no sense? I don't see the point of checking for errors here tbh (at least for now, maybe it makes more sense on subsequent `Save` calls.)

Reply:
It's defensive programming in this case, because we don't want to ever silently do something that might implicitly break the core business logic.